src/pasta/hashtocurve.rs

```diff
@@ -0,0 +1,145 @@
+//! This module implements "simplified SWU" hashing to short Weierstrass curves
+//! with a = 0.
+
+use subtle::ConstantTimeEq;
+
+use crate::arithmetic::{Curve, CurveAffine, FieldExt};
+
+/// Hashes over a message and writes the output to all of `buf`.
+pub fn hash_to_field<F: FieldExt>(message: &[u8], domain_separation_tag: &[u8], buf: &mut [F]) {
+    use sha3::digest::{ExtendableOutput, Update};
+    assert!(domain_separation_tag.len() < 256);
+
+    // Assume that the field size is 32 bytes and k is 256, where k is defined in
+    // <https://www.ietf.org/archive/id/draft-irtf-cfrg-hash-to-curve-10.html#name-security-considerations-3>.
+    const CHUNKLEN: usize = 64;
+
+    let outlen = buf.len() * CHUNKLEN;
+
+    let mut xof = sha3::Shake128::default();
+    xof.update(message);
+    xof.update(&(outlen as u32).to_be_bytes());
+    xof.update([domain_separation_tag.len() as u8]);
+    xof.update(domain_separation_tag);
+
+    for (big, buf) in xof
+        .finalize_boxed(outlen)
+        .chunks(CHUNKLEN)
+        .zip(buf.iter_mut())
+    {
+        let mut little = [0u8; CHUNKLEN];
+        little.copy_from_slice(big);
+        little.reverse();
+        *buf = F::from_bytes_wide(&little);
+    }
+}
+
+/// Implements a degree 3 isogeny map.
+pub fn iso_map<F: FieldExt, C: CurveAffine<Base = F>, I: CurveAffine<Base = F>>(
+    p: &I::Projective,
+    iso: &[C::Base; 13],
+) -> C::Projective {
+    // The input and output are in Jacobian coordinates, using the method
+    // in "Avoiding inversions" [WB2019, section 4.3].
+
+    let (x, y, z) = p.jacobian_coordinates();
+
+    let z2 = z.square();
+    let z3 = z2 * z;
+    let z4 = z2.square();
+    let z6 = z3.square();
+
+    let num_x = ((iso[0] * x + iso[1] * z2) * x + iso[2] * z4) * x + iso[3] * z6;
+    let div_x = (z2 * x + iso[4] * z4) * x + iso[5] * z6;
+
+    let num_y = (((iso[6] * x + iso[7] * z2) * x + iso[8] * z4) * x + iso[9] * z6) * y;
+    let div_y = (((x + iso[10] * z2) * x + iso[11] * z4) * x + iso[12] * z6) * z3;
+
+    let zo = div_x * div_y;
+    let xo = num_x * div_y * zo;
+    let yo = num_y * div_x * zo.square();
+
+    C::Projective::new_jacobian(xo, yo, zo).unwrap()
+}
+
+pub fn map_to_curve<F: FieldExt, C: CurveAffine<Base = F>, I: CurveAffine<Base = F>>(
```

Comment:
should we rename this `map_to_curve_simple_swu` instead? as per https://www.ietf.org/archive/id/draft-irtf-cfrg-hash-to-curve-10.html#name-simplified-swu-for-ab-0-2